dbms/src/TiDB/Etcd/Client.cpp

```diff
@@ -0,0 +1,270 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Poco/URI.h>
+#include <TiDB/Etcd/Client.h>
+#include <TiDB/Etcd/EtcdConnClient.h>
+#include <common/logger_useful.h>
+#include <etcd/rpc.grpc.pb.h>
+#include <etcd/rpc.pb.h>
+#include <etcd/v3election.grpc.pb.h>
+#include <etcd/v3election.pb.h>
+#include <fmt/chrono.h>
+
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#endif
+#include <grpcpp/client_context.h>
+#include <grpcpp/impl/codegen/call_op_set.h>
+#include <grpcpp/support/status.h>
+#include <grpcpp/support/status_code_enum.h>
+#ifdef __clang__
+#pragma clang diagnostic pop
+#endif
+
+#include <chrono>
+#include <mutex>
+
+namespace DB::Etcd
+{
+ClientPtr Client::create(const pingcap::pd::ClientPtr & pd_client, const pingcap::ClusterConfig & config, Int64 timeout_s)
+{
+    auto etcd_client = std::make_shared<Client>();
+    etcd_client->pd_client = pd_client;
+    etcd_client->timeout = std::chrono::seconds(timeout_s);
+    etcd_client->config = config;
+    etcd_client->log = Logger::get();
+
+    return etcd_client;
+}
+
+void Client::update(const pingcap::ClusterConfig & new_config)
+{
+    // Note that config of pd_client is updated outside of this function
+    // before the etcd client config update.
+
+    {
+        std::lock_guard lk(mtx_channel_map);
+        config = new_config;
+        channel_map.clear();
+    }
+}
+
+EtcdConnClientPtr Client::getOrCreateGRPCConn(const String & addr)
+{
+    std::lock_guard lk(mtx_channel_map);
+    auto it = channel_map.find(addr);
+    if (it != channel_map.end())
+        return it->second;
+
+    Poco::URI uri(addr);
+    auto client = std::make_shared<EtcdConnClient>(uri.getAuthority(), config);
```

Comment:
construct `client` before acquiring lock?

Reply:
No. We try to reuse the connection under the lock. Only when there is no connection to `addr`, do we create a new grpc connection to etcd.

Reply:
It can reduce the time interval of mutually exclusive lock, a trade-off.